client/src/controllers/ProgressController.ts

```diff
@@ -61,8 +61,25 @@ export class ProgressController extends Controller {
             `.${DEFAULT_CLASS}:not([${controllerAttribute}~='${identifier}'])`,
           )
           .forEach((button) => {
-            button.setAttribute(controllerAttribute, identifier);
-            button.setAttribute(actionAttribute, `${identifier}#activate`);
+            // If there is an existing controller and action attribute on the element
+            let existingControllerValue =
+              button.getAttribute(controllerAttribute);
+            let existingActionValue = button.getAttribute(actionAttribute);
+
+            if (existingControllerValue) {
+              existingControllerValue = `${existingControllerValue} ${identifier}`;
+            } else {
+              existingControllerValue = identifier;
+            }
+
+            if (existingActionValue) {
+              existingActionValue = `${existingActionValue} ${identifier}#activate`;
+            } else {
+              existingActionValue = `${identifier}#activate`;
+            }
+
+            button.setAttribute(controllerAttribute, existingControllerValue);
+            button.setAttribute(actionAttribute, existingActionValue);
```

Comment:
Just a quick suggestion - mainly, try to keep it simple and avoid building up the same string one at a time. Additionally, do one 'thing' at a time, to the controller attribute, and then do the action attribute, try not to group together.

```suggestion
            // set the controller attribute, appending to existing if present
            button.setAttribute(
              controllerAttribute,
              [button.getAttribute(controllerAttribute) || '', identifier]
                .filter(Boolean)
                .join(' '),
            );

            // set the action attribute, appending to existing if present
            button.setAttribute(
              actionAttribute,
              [
                button.getAttribute(actionAttribute) || '',
                `${identifier}#activate`,
              ]
                .filter(Boolean)
                .join(' '),
            );

            // set the active text label to replace the legacy data-click-text
```

Also includes a comment for the next 'block' of logic.